tests/test_lion8b.py

```diff
@@ -0,0 +1,422 @@
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('N,D', [(1, 8), (17, 23), (32, 32)])
+@pytest.mark.parametrize('device', ['cpu', 'cuda'])
+@pytest.mark.parametrize('dtype', _FLOAT_DTYPES)
+@pytest.mark.parametrize('fused,use_errors', [(False, False), (True, False),
+                                              (True, True)])
+def test_descends(N: int, D: int, device: str, dtype: torch.dtype, fused: bool,
+                  use_errors: bool) -> None:
+    if (dtype != torch.float32) and device == 'cpu':
+        return
+    torch.manual_seed(123)
+    X = torch.randn((N, D), device=device, requires_grad=False, dtype=dtype)
+    W = torch.randn((D, D), device=device, requires_grad=True, dtype=dtype)
+
+    # we use tiny beta1 so we move almost entirely in the gradient direction
+    opt = Lion8bit([W],
+                   lr=1e-2,
+                   betas=(.5, .5),
+                   quantize=(device != 'cpu'),
+                   _fused=fused,
+                   error_correction=use_errors)
+
+    prev_loss = np.inf
+    prev_momentum = None
+    num_iters = 10 if device == 'cuda' else 2  # keep test fast
+    for _ in range(num_iters):
+        Y = X @ W
+        loss = (Y * Y).mean()
+        loss.backward()
+        opt.step()
+        opt.zero_grad()
+
+        loss_val = loss.item()
+        assert loss_val < prev_loss
+        prev_loss = loss_val
+
+        # since we're getting the same batch every time and have a small
+        # learning rate, our gradients should point in the same direction
+        # at each step. Consequently, our momentum should grow each step.
+        state_for_param = opt.state[W]  # type:ignore using tensor as key
+        momentum = state_for_param['exp_avg'].materialize()
+        assert momentum is not None and momentum.shape == W.shape
+        if prev_momentum is not None:
+            momentum_changes = momentum - prev_momentum
+            assert torch.all(momentum_changes >= 0)
+            assert momentum_changes.max() > 0
+            prev_momentum = momentum
+
+
+def _nmse(vals_true: torch.Tensor,
+          vals_hat: torch.Tensor,
+          norm_how: str = 'l2_sq'):
+    diffs = vals_true - vals_hat
+    mse = (diffs * diffs).mean()
+    if norm_how == 'var':
+        return mse / vals_true.var()
+    return mse / (vals_true * vals_true).mean()
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('w_init', ['cyclic', 'rand'])
+@pytest.mark.parametrize('grad_strategy', ['zero', 'ones', 'const', 'rand'])
+@pytest.mark.parametrize('D', [4, 12])  # vectorized and unvectorized impls
+@pytest.mark.parametrize('dtype', _FLOAT_DTYPES)
+def test_lion8b_fused_unfused_unquantized_same(w_init: str, grad_strategy: str,
+                                               D: int,
+                                               dtype: torch.dtype) -> None:
+    torch.manual_seed(123)
+    device = 'cuda'
+
+    # each optimizer gets a different weight matrix to optimize
+    if w_init == 'cyclic':
```

Comment:
is this a particularly bad case?
why test this init?

Reply:
mostly because it was good for debugging. I can rm it.